// @sudoku/stores/puzzle-grid.js
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { generateSudoku } from '@sudoku/sudoku';
import { writable } from 'svelte/store';
import { SUDOKU_SIZE } from '@sudoku/constants';

/**
 * 题目网格管理器 - 只负责原始题目的管理
 * 
 * OOAD设计原则：
 * 1. 单一职责：只负责题目网格的存储和管理
 * 2. 开闭原则：可以扩展新的题目来源而不修改现有代码
 */
function createPuzzleGrid() {
	const grid = writable(Array(SUDOKU_SIZE).fill(null).map(() => Array(SUDOKU_SIZE).fill(0)));

	return {
		subscribe: grid.subscribe,

		/**
		 * 生成新题目
		 * @param {('veryeasy' | 'easy' | 'medium' | 'hard')} difficulty - 难度
		 */
		generate(difficulty) {
			grid.set(generateSudoku(difficulty));
		},

		/**
		 * 从编码字符串解码题目
		 * @param {string} sencode - 编码字符串
		 */
		decodeSencode(sencode) {
			grid.set(decodeSencode(sencode));
		},

		/**
		 * 获取指定位置的值
		 */
		get(gridStore, x, y) {
			return gridStore[y][x];
		},

		/**
		 * 获取编码字符串
		 */
		getSencode(gridStore) {
			return encodeSudoku(gridStore);
		},
	};
}

export const puzzleGrid = createPuzzleGrid();
