// @sudoku/strategies/strategy-registry.js
import { HiddenSingleStrategy, NakedPairsStrategy, PointingPairsStrategy, BoxLineReductionStrategy, NakedTriplesStrategy } from './basic-strategies.js';
import { 
	BUGStrategy, XWingStrategy, UniqueRectangles1Strategy, ChuteRemotePairsStrategy, WWingStrategy,
	YWingStrategy, RectangleEliminationStrategy, SimpleColouringStrategy, QuadsStrategy, SwordfishStrategy,
	XYZWingStrategy, XCyclesStrategy, XYChainStrategy, Medusa3DStrategy, JellyfishStrategy,
	UniqueRectangles2PlusStrategy, VariousStaticPatternsStrategy, ExtendedUniqueRectsStrategy,
	HiddenUniqueRectsStrategy, WXYZWingStrategy, AlignedPairExclusionStrategy, ExocetStrategy,
	GroupedXCyclesStrategy, FinnedXWingStrategy, FinnedSwordfishStrategy, AlternInferenceChainsStrategy,
	AlmostLockedSetsStrategy, DeathBlossomStrategy, SudeCoqStrategy, DigitForcingChainsStrategy,
	NishioForcingChainsStrategy, CellForcingChainsStrategy, UnitForcingChainsStrategy,
	PatternOverlayMethodStrategy, QuadForcingChainsStrategy, BowmansBingoStrategy
} from './advanced-strategies.js';
import { BaseStrategy } from './base-strategy.js';

/**
 * 策略注册表 - 管理所有数独求解策略
 * 
 * OOAD设计原则：
 * 1. 单一职责：负责策略的注册和管理
 * 2. 开闭原则：可以动态添加新策略而不修改现有代码
 * 3. 依赖注入：策略通过注册表统一管理
 */
class StrategyRegistry {
	constructor() {
		this.strategies = new Map();
		this.registerDefaultStrategies();
	}

	/**
	 * 注册默认策略（基础策略1-5）
	 */
	registerDefaultStrategies() {
		// 基础策略 1-5
		this.register(new HiddenSingleStrategy());
		this.register(new NakedPairsStrategy());
		this.register(new PointingPairsStrategy());
		this.register(new BoxLineReductionStrategy());
		this.register(new NakedTriplesStrategy());

		// Tough策略 6-16
		this.register(new BUGStrategy());
		this.register(new XWingStrategy());
		this.register(new UniqueRectangles1Strategy());
		this.register(new ChuteRemotePairsStrategy());
		this.register(new WWingStrategy());
		this.register(new YWingStrategy());
		this.register(new RectangleEliminationStrategy());
		this.register(new SimpleColouringStrategy());
		this.register(new QuadsStrategy());
		this.register(new SwordfishStrategy());
		this.register(new XYZWingStrategy());

		// Diabolical策略 17-26
		this.register(new XCyclesStrategy());
		this.register(new XYChainStrategy());
		this.register(new Medusa3DStrategy());
		this.register(new JellyfishStrategy());
		this.register(new UniqueRectangles2PlusStrategy());
		this.register(new VariousStaticPatternsStrategy());
		this.register(new ExtendedUniqueRectsStrategy());
		this.register(new HiddenUniqueRectsStrategy());
		this.register(new WXYZWingStrategy());
		this.register(new AlignedPairExclusionStrategy());

		// Extreme策略 27-40
		this.register(new ExocetStrategy());
		this.register(new GroupedXCyclesStrategy());
		this.register(new FinnedXWingStrategy());
		this.register(new FinnedSwordfishStrategy());
		this.register(new AlternInferenceChainsStrategy());
		this.register(new AlmostLockedSetsStrategy());
		this.register(new DeathBlossomStrategy());
		this.register(new SudeCoqStrategy());
		this.register(new DigitForcingChainsStrategy());
		this.register(new NishioForcingChainsStrategy());
		this.register(new CellForcingChainsStrategy());
		this.register(new UnitForcingChainsStrategy());
		this.register(new PatternOverlayMethodStrategy());
		this.register(new QuadForcingChainsStrategy());

		// Trial and Error策略 41
		this.register(new BowmansBingoStrategy());
	}

	/**
	 * 注册策略
	 */
	register(strategy) {
		if (!(strategy instanceof BaseStrategy)) {
			throw new Error('Strategy must extend BaseStrategy');
		}
		this.strategies.set(strategy.id, strategy);
	}

	/**
	 * 获取策略
	 */
	get(id) {
		return this.strategies.get(id);
	}

	/**
	 * 获取所有策略（按ID排序）
	 */
	getAll() {
		return Array.from(this.strategies.values()).sort((a, b) => a.id - b.id);
	}

	/**
	 * 获取指定范围的策略
	 */
	getRange(startId, endId) {
		const result = [];
		for (let id = startId; id <= endId; id++) {
			const strategy = this.strategies.get(id);
			if (strategy) result.push(strategy);
		}
		return result;
	}
}

/**
 * 占位策略 - 用于尚未实现的策略
 * 确保框架完整性，新策略可以逐步替换占位策略
 */
class PlaceholderStrategy extends BaseStrategy {
	constructor(id) {
		const names = {
			6: 'BUG', 7: 'X-Wing', 8: 'Unique Rectangles 1', 9: 'Chute Remote Pairs',
			10: 'W-Wing', 11: 'Y-Wing', 12: 'Rectangle Elimination', 13: 'Simple Colouring',
			14: 'Quads Naked+Hidden', 15: 'Swordfish', 16: 'XYZ Wing',
			17: 'X-Cycles', 18: 'XY-Chain', 19: '3D Medusa', 20: 'Jellyfish',
			21: 'Unique Rectangles 2+', 22: 'Various static patterns', 23: 'Extended Unique Rect\'s',
			24: 'Hidden Unique Rect\'s', 25: 'WXYZ Wing', 26: 'Aligned Pair Exclusion',
			27: 'Exocet', 28: 'Grouped X-Cycles', 29: 'Finned X-Wing', 30: 'Finned Swordfish',
			31: 'Altern. Inference Chains', 32: 'Almost Locked Sets', 33: 'Death Blossom',
			34: 'Sue-de-Coq', 35: 'Digit Forcing Chains', 36: 'Nishio Forcing Chains',
			37: 'Cell Forcing Chains', 38: 'Unit Forcing Chains', 39: 'Pattern Overlay Method',
			40: 'Quad Forcing Chains', 41: 'Bowman\'s Bingo',
		};

		const categories = {
			6: 'Tough', 7: 'Tough', 8: 'Tough', 9: 'Tough', 10: 'Tough', 11: 'Tough',
			12: 'Tough', 13: 'Tough', 14: 'Tough', 15: 'Tough', 16: 'Tough',
			17: 'Diabolical', 18: 'Diabolical', 19: 'Diabolical', 20: 'Diabolical',
			21: 'Diabolical', 22: 'Diabolical', 23: 'Diabolical', 24: 'Diabolical',
			25: 'Diabolical', 26: 'Diabolical',
			27: 'Extreme', 28: 'Extreme', 29: 'Extreme', 30: 'Extreme', 31: 'Extreme',
			32: 'Extreme', 33: 'Extreme', 34: 'Extreme', 35: 'Extreme', 36: 'Extreme',
			37: 'Extreme', 38: 'Extreme', 39: 'Extreme', 40: 'Extreme',
			41: 'Trial and Error',
		};

		super(names[id] || `Strategy ${id}`, id, categories[id] || 'Unknown');
	}

	execute(grid) {
		// 占位策略返回null，表示尚未实现
		return null;
	}
}

export const strategyRegistry = new StrategyRegistry();
