// @sudoku/services/game-state-snapshot.js
import { get } from 'svelte/store';
import { userGrid } from '../stores/grid.js';
import { candidates } from '../stores/candidates.js';
import { hints, usedHints } from '../stores/hints.js';

/**
 * 游戏状态快照 - 负责创建和管理游戏状态的快照
 * 
 * OOAD设计原则：
 * 1. 单一职责：专门负责状态快照的创建和克隆
 * 2. 依赖倒置：通过接口访问状态，不直接依赖具体实现
 */
class GameStateSnapshot {
	/**
	 * 创建当前游戏状态的快照
	 * @returns {Object} 状态快照 {grid, candidates, hints}
	 */
	create() {
		return {
			grid: this.cloneGrid(get(userGrid)),
			candidates: this.cloneCandidates(get(candidates)),
			hints: {
				remaining: get(hints),
				used: get(usedHints),
			},
		};
	}

	/**
	 * 应用状态快照
	 * @param {Object} snapshot - 状态快照
	 */
	apply(snapshot) {
		userGrid.replace(snapshot.grid);
		candidates.replace(snapshot.candidates);
		hints.restore(snapshot.hints.remaining, snapshot.hints.used);
	}

	/**
	 * 克隆网格
	 */
	cloneGrid(grid) {
		return grid.map(row => [...row]);
	}

	/**
	 * 克隆候选数
	 */
	cloneCandidates(currentCandidates) {
		const cloned = {};
		Object.entries(currentCandidates || {}).forEach(([key, values]) => {
			cloned[key] = [...values];
		});
		return cloned;
	}

	/**
	 * 比较两个状态是否相等
	 */
	areEqual(a, b) {
		return JSON.stringify(a) === JSON.stringify(b);
	}
}

export const gameStateSnapshot = new GameStateSnapshot();
