// @sudoku/services/history-manager.js
import { writable } from 'svelte/store';
import { gameStateSnapshot } from './game-state-snapshot.js';

/**
 * 历史管理器 - 负责管理游戏的撤销/重做历史
 * 
 * OOAD设计原则：
 * 1. 单一职责：专门负责历史管理
 * 2. 依赖倒置：依赖GameStateSnapshot抽象，不直接操作store
 * 3. 开闭原则：可以扩展新的历史管理策略而不修改现有代码
 */
class HistoryManager {
	constructor(snapshotService = gameStateSnapshot) {
		this.snapshotService = snapshotService;
		this.undoStack = [];
		this.redoStack = [];
		this.status = writable({ canUndo: false, canRedo: false });
	}

	subscribe(run) {
		return this.status.subscribe(run);
	}

	reset() {
		this.undoStack = [this.snapshotService.create()];
		this.redoStack = [];
		this.updateStatus();
	}

	applyChange(mutator) {
		const before = this.snapshotService.create();
		mutator();
		const after = this.snapshotService.create();

		if (this.snapshotService.areEqual(before, after)) return;

		this.undoStack.push(before);
		this.redoStack = [];
		this.updateStatus();
	}

	undo() {
		if (this.undoStack.length <= 1) return;

		const current = this.snapshotService.create();
		const previous = this.undoStack[this.undoStack.length - 2];

		this.redoStack.push(current);
		this.undoStack.pop();

		this.snapshotService.apply(previous);
		this.updateStatus();
	}

	redo() {
		if (this.redoStack.length === 0) return;

		const current = this.snapshotService.create();
		const next = this.redoStack.pop();

		this.undoStack.push(current);

		this.snapshotService.apply(next);
		this.updateStatus();
	}

	updateStatus() {
		this.status.set({
			canUndo: this.undoStack.length > 1,
			canRedo: this.redoStack.length > 0,
		});
	}
}

export const history = new HistoryManager();
export const historyState = history.status;
