// @sudoku/services/strategy-manager.js
import { strategyRegistry } from '../strategies/strategy-registry.js';
import { strategyValidator } from './strategy-validator.js';

/**
 * 策略管理器 - 按顺序执行策略并收集结果
 * 
 * OOAD设计原则：
 * 1. 单一职责：负责策略的执行和结果管理
 * 2. 策略模式：通过策略接口统一调用不同策略
 * 3. 模板方法：定义策略执行的统一流程
 * 4. 观察者模式：可以扩展为支持策略执行事件通知
 */
class StrategyManager {
	constructor() {
		this.strategies = strategyRegistry.getAll();
	}

	/**
	 * 获取所有策略（供外部使用）
	 */
	getAllStrategies() {
		return this.strategies;
	}

	/**
	 * 执行所有策略，找到第一个可用的策略
	 * 
	 * @param {number[][]} grid - 数独网格
	 * @returns {Object} 执行结果
	 * {
	 *   success: boolean,
	 *   result: Object|null,  // 第一个成功的策略结果
	 *   tested: Array,        // 所有测试过的策略
	 *   failed: Array,        // 测试过但不可用的策略
	 *   successful: Array,     // 可用的策略
	 *   untested: Array,       // 未验证的策略（找到第一个成功后停止）
	 *   usedStrategy: Object|null  // 实际采用的策略
	 * }
	 */
	executeStrategies(grid) {
		const tested = [];
		const failed = [];
		const successful = [];
		const untested = [];
		let firstResult = null;
		let usedStrategy = null;

		// 按顺序执行所有策略
		for (const strategy of this.strategies) {
			const validation = strategyValidator.validateStrategy(grid, strategy);

			const testResult = {
				strategy: strategy.getDescription(),
				applicable: validation.applicable || false,
				valid: validation.valid,
				reason: validation.reason || null,
				result: validation.result || null,
			};

			tested.push(testResult);

			if (!validation.valid) {
				failed.push(testResult);
				continue;
			}

			if (validation.applicable && validation.result) {
				successful.push(testResult);

				// 记录第一个成功的结果
				if (!firstResult) {
					firstResult = validation.result;
					usedStrategy = testResult;
					// 找到第一个成功后，记录剩余未验证的策略
					const currentIndex = this.strategies.indexOf(strategy);
					for (let i = currentIndex + 1; i < this.strategies.length; i++) {
						untested.push({
							strategy: this.strategies[i].getDescription(),
						});
					}
					break; // 找到第一个成功后停止
				}
			} else {
				// 策略有效但不适用当前网格
				failed.push(testResult);
			}
		}

		// 如果没有找到成功的策略，所有策略都是tested或failed
		if (!firstResult) {
			// 所有策略都已测试，没有未验证的
		}

		return {
			success: firstResult !== null,
			result: firstResult,
			tested,
			failed,
			successful,
			untested,
			usedStrategy,
		};
	}

	/**
	 * 执行指定范围的策略
	 */
	executeStrategyRange(grid, startId, endId) {
		const strategies = strategyRegistry.getRange(startId, endId);
		const tested = [];
		const failed = [];
		const successful = [];
		let firstResult = null;

		for (const strategy of strategies) {
			const validation = strategyValidator.validateStrategy(grid, strategy);

			const testResult = {
				strategy: strategy.getDescription(),
				applicable: validation.applicable || false,
				valid: validation.valid,
				reason: validation.reason || null,
				result: validation.result || null,
			};

			tested.push(testResult);

			if (!validation.valid) {
				failed.push(testResult);
				continue;
			}

			if (validation.applicable && validation.result) {
				successful.push(testResult);
				if (!firstResult) {
					firstResult = validation.result;
				}
			} else {
				failed.push(testResult);
			}
		}

		return {
			success: firstResult !== null,
			result: firstResult,
			tested,
			failed,
			successful,
		};
	}

	/**
	 * 获取策略统计信息
	 */
	getStatistics() {
		const stats = {
			total: this.strategies.length,
			byCategory: {},
		};

		for (const strategy of this.strategies) {
			const category = strategy.category;
			if (!stats.byCategory[category]) {
				stats.byCategory[category] = 0;
			}
			stats.byCategory[category]++;
		}

		return stats;
	}
}

export const strategyManager = new StrategyManager();
